File: lib/codetogether/games/game.ex
defmodule Codetogether.Games.Game do
  use Ecto.Schema
  import Ecto.Changeset

  schema "games" do
    field :can_vote, :boolean, default: false
    field :complete, :boolean, default: false
    field :confetti, :boolean, default: false
    field :description, :string
    field :max_input, :integer
    field :mode, :string
    field :name, :string
    field :play_state, :string
    field :vote_interval, :integer
    field :vote_threshold, :integer

    has_many :messages, Codetogether.Games.Message
    has_many :entries, Codetogether.Games.Entry

    timestamps()
  end

  @doc false
  def changeset(game, attrs) do
    game
    |> cast(attrs, [
      :name,
      :description,
      :mode,
      :complete,
      :max_input,
      :vote_interval,
      :vote_threshold,
      :can_vote,
      :confetti,
      :play_state
    ])
    |> validate_required([
      :name,
      :description,
      :mode,
      :complete,
      :max_input,
      :vote_interval,
      :vote_threshold,
      :can_vote,
      :confetti,
      :play_state
    ])
  end
end

File: assets/js/Game.ts
interface Nominee {
  body: string;
  id: number;
  vote_count: number;
}

interface Game {
  can_vote: boolean;
  code: string;
  complete: boolean;
  confetti: boolean;
  description: string;
  id: number;
  inserted_at: Date;
  max_input: number;
  messages: Message[];
  mode: string;
  name: string;
  nominees: Nominee[];
  output: { raw: string; err_ln: number };
  play_state: string;
  updated_at: Date;
  vote_interval: number;
  vote_threshold: number;
}

export type Message = {
  id: number;
  name: string;
  color: string;
  is_code: boolean;
  user_id: number;
};
export type { Game };
